pkp/pkp-lib#11589 Consider JAV versions for indexing

bozana opened the pull request:
s. https://github.com/pkp/pkp-lib/issues/11589

this PR introduces:
- a new status depositable (a help status), stale, submitted (when jobs that handle registration is dispatched), and error
- listening and acting to the new hooks on publication version, publish and unpublish (as described in the issue)
- new publications listing/grid (similar to the submission's one) for export plugins to be used
- possibility to see error messages, if the registration has failed
- some code refactoring

(It would be best to review the changes thinking about and considering the requirements from DOAJ and Zenodo plugin, as two examples. Maybe there will need to be some adjustments once these plugins are fully implemented and the publications export and deposit is considered there as well, but this should be a good bases, as far as I could think about it.)

bozana commented:
Hi @kaitlinnewson, yes, it needs to be used/implemented in export plugins. Here an example, how to use it in the DOAJ plugin: https://github.com/pkp/ojs/pull/4986/files#diff-15d46d2256d0455db8ef22e87c303f713edb68c63313e8e22b65689fdbc9b1bc.
Thanks a lot!

kaitlinnewson commented on classes/plugins/PubObjectsExportPlugin.php:
```diff
@@ -797,6 +919,23 @@ public function getPublishedSubmissions($submissionIds, $context)
+        $allPublicationIds = Repo::publication()
+            ->getCollector()
+            ->filterByContextIds([$context->getId()])
+            ->filterByStatus([PKPSubmission::STATUS_PUBLISHED])
```
Will this work? It doesn't look like the publication collector has this filter.

asmecher commented on classes/plugins/DOIPubIdExportPlugin.php:
```diff
@@ -69,17 +69,22 @@ public function getPubIdDisplayType()
+            if ($object instanceof Submission) {
+                $doiIds = Repo::doi()->getDoisForSubmission($object->getId());
+            } else {
+                $doiIds = Repo::doi()->getDoisForIssue($object->getId, true);
```
:up: on that! I've been using `match` for this kind of thing lately:
```php
$doiIds = match(true) {
    $object instanceof Submission => Repo::doi()->getDoisForSubmission($object->getId()),
    $object instanceof Issue => Repo::doi()->getDoisForIssue($object->getId, true),
};
```
`match` statements have the benefit of automatically throwing an `UnhandledMatchError` if no case matches, which is generally what we want at no added lines of code.

kaitlinnewson commented:
In testing this with the DOAJ changes in https://github.com/pkp/ojs/pull/4986, it looks like there may be an issue with the pagination counts for the publications to look into:

<img width="971" height="627" alt="Screenshot 2025-08-13 at 11 25 26 AM" src="https://github.com/user-attachments/assets/74d97437-1b51-4404-bbf9-b5a02b83fef8" />

kaitlinnewson commented on classes/publication/DAO.php:
```diff
@@ -56,4 +65,77 @@ public function fromRow(object $primaryRow): Publication
+                    )
+                )
+            )
+            ->orderByDesc('s.submission_id')
```
Could this be missing a groupBy, e.g. `->groupBy('p.publication_id')`? It seems to fix the count and display issue on the DOAJ plugin as well.

bozana commented:
Hi @kaitlinnewson, I considered you comments, and additionally:
- fixed handlePublicationVersioning, handlePublicationPublishing, handlePublicationUnpublishing
- fixed SQL in publication DAO
- considered additional help constant EXPORT_STATUS_DEPOSITABLE, used to get all depositable submissions/publications, e.g. from a scheduled task
- added constant EXPORT_STATUS_ERROR, as well as the possibility to save the request failed message, and display it in a modal
- slightly reordered the functions in PubObjectsExportPlugin

Could you please take another look? Above all the changes in the PubObjectsExportPlugin.
Thanks a lot!

kaitlinnewson commented on controllers/grid/publications/ExportPublishedPublicationsListGridHandler.php:
```diff
@@ -0,0 +1,318 @@
+
+        // Allow for plugins that have already been loaded (e.g. by injection from a generic plugin)
+        // Otherwise, load the plugin as specified
+        $this->_plugin = PluginRegistry::getPlugin($pluginCategory, $pluginPathName) ?? PluginRegistry::loadPlugin($pluginCategory, $pluginPathName);
```
When using the search feature in the DOAJ plugin, an error happens around here and it doesn't seem to load the plugin (maybe due to the importexport category vs the plugin location?): 
```
PHP Fatal error:  Uncaught TypeError: Cannot assign null to property APP\controllers\grid\publications\ExportPublishedPublicationsListGridHandler::$_plugin of type APP\plugins\PubObjectsExportPlugin in /Users/kaitlinnewson/Code/ojs-main/controllers/grid/publications/ExportPublishedPublicationsListGridHandler.php:84
```

bozana replied:
Oh, yes, moving it to the generic plugin changes a few things...

kaitlinnewson commented on plugins/generic/doaj/jobs/DOAJDelete.php:
```diff
@@ -0,0 +1,59 @@
+            $msg = __($result[0][0], ['param' => $result[0][1]]);
+            throw new JobException($msg);
+        }
+        dispatch(new DOAJRegister($this->jsonString, $this->objectId, $this->context))->delay(now()->addMinutes(10));
```
In the case where a metadata update happens, and we delete then register the updated record, does the status ever get updated from "submitted" to "registered"?

kaitlinnewson replied:
I noticed mine seemed to remain in the "submitted" status in the case of the deletion needed, but I'll take a closer look in case I missed something.

bozana replied:
You would need to wait 10 minutes, and access a page, for the register job to be run then

kaitlinnewson commented on classes/plugins/PubObjectsExportPlugin.php:
```diff
@@ -397,33 +411,19 @@ public function getRepresentationFilter()
+     * @return array (string action => string text)
      */
-    public function getStatusNames()
+    public function getExportActionNames()
```
Add return type? (And in DOAJExportPlugin)

bozana replied:
This is actually the old function, I just resorted the functions, so I have not change anything, but I can add the return type -- it seems no other plugin uses it currently...

kaitlinnewson commented on plugins/generic/doaj/DOAJExportPlugin.php:
```diff
@@ -0,0 +1,381 @@
+                    'query' => ['api_key' => $apiKey, 'article_id' => $doajId],
+                ]
+            );
+        } catch (\GuzzleHttp\Exception\RequestException $e) {
```
Add GuzzleException as well?

bozana commented on classes/plugins/PubObjectsExportGenericPlugin.php:
```diff
@@ -0,0 +1,191 @@
+                ->filterByVersionMajor($newPublication->getData('versionMajor'))
+                ->filterByStatus([PKPSubmission::STATUS_PUBLISHED])
+                ->getMany()
+                ->last(); // minor versions are sorted ASC, so get only the last
```
Correct, the publications are per default ordered so in the submission object, but here we get them differently...

bozana commented on classes/plugins/PubObjectsExportPlugin.php:
```diff
@@ -576,12 +667,10 @@ public function addToSchema($hookName, $params)
+        return [$this->getDepositStatusSettingName(), $this->getFailedMsgSettingName()];
     }
 
     /**
```
I will try and see...

bozana replied:
Yes, this seems to work

bozana commented on plugins/generic/doaj/DOAJExportPlugin.php:
```diff
@@ -0,0 +1,381 @@
+            if (empty($resultErrors)) {
+                $this->_sendNotification(
+                    $request->getUser(),
+                    $this->getDepositSuccessNotificationMessageKey(),
```
I will then use "Articles submitted successfully"

kaitlinnewson commented on plugins/generic/doaj/DOAJExportPlugin.php:
```diff
@@ -0,0 +1,405 @@
+            // redirect back to the right tab
+            $request->redirect(null, null, null, $path, null, $tab);
+        } else {
+            return parent::executeExportAction($request, $objects, $filter, $tab, $objectsFileNamePart, $noValidation);
```
```suggestion
           parent::executeExportAction($request, $objects, $filter, $tab, $objectsFileNamePart, $noValidation);
```

kaitlinnewson commented on plugins/generic/doaj/filter/DOAJJsonFilter.php:
```diff
@@ -60,31 +62,53 @@ public function &process(&$pubObject)
+        }
         $publicationLocale = $publication->getData('locale');
 
         $issueId = $publication->getData('issueId');
```
Could also initialize $issue = null; around here to reduce warnings later.

kaitlinnewson commented on plugins/generic/doaj/DOAJExportPlugin.php:
```diff
@@ -0,0 +1,405 @@
+    {
+        $filterDao = DAORegistry::getDAO('FilterDAO'); /** @var FilterDAO $filterDao */
+        $exportFilters = $filterDao->getObjectsByGroup($filter);
+        assert(count($exportFilters) == 1); // Assert only a single serialization filter
```
Is this needed? Maybe convert to exception if so?

kaitlinnewson commented on plugins/generic/doaj/DOAJInfoSender.php:
```diff
@@ -125,9 +129,8 @@ public function _registerObjects($objects, $filter, $journal, $objectsFileNamePa
-     * @param array $errors
      */
-    public function _addLogEntry($errors)
+    protected function _addLogEntry(array $errors): void
```
Could convert `assert` here to exception if needed?

kaitlinnewson commented on classes/publication/DAO.php:
```diff
@@ -55,4 +64,81 @@ public function fromRow(object $primaryRow): Publication
+            ->whereNull('p2.publication_id')
+            ->when($pubIdType != null, fn (Builder $q) => $q->where('pspidt.setting_name', '=', "pub-id::{$pubIdType}")->whereNotNull('pspidt.setting_value'))
+            ->when($title != null, fn (Builder $q) => $q->where('pst.setting_name', '=', 'title')->where('pst.setting_value', 'LIKE', "%{$title}%"))
+            ->when($author != null, fn (Builder $q) => $q->whereRaw("CONCAT(COALESCE(asgs.setting_value, ''), ' ', COALESCE(asfs.setting_value, '')) LIKE ?", [$author]))
```
Could adjust this to allow for only last name/first name searches (e.g. %author% or something similar)

bozana replied:
if we allow this only for last and first name, should we change it in the UI as well to be "Author name" instead of "Author"?

bozana replied:
What do you exactly mean? To allow searches with only partial match (%author%)?

kaitlinnewson replied:
I was thinking for example if you were to search for only the last name, to handle that case.

bozana replied:
Ah yes, good point!

kaitlinnewson commented on classes/plugins/PubObjectsExportPlugin.php:
```diff
@@ -477,7 +534,7 @@ public function exportXML($objects, $filter, $context, $noValidation = null, &$o
         $exportFilters = $filterDao->getObjectsByGroup($filter);
         assert(count($exportFilters) == 1); // Assert only a single serialization filter
-        $exportFilter = array_shift($exportFilters);
+        $exportFilter = array_shift($exportFilters); /** @var PKPImportExportFilter $exportFilter */
```
Import needed for `PKPImportExportFilter`

bozana commented on plugins/generic/doaj/filter/DOAJJsonFilter.php:
```diff
@@ -60,31 +62,53 @@ public function &process(&$pubObject)
+                $article = $cache->get('articles', $publication->getData('submissionId'));
+            } else {
+                $article = Repo::submission()->get($publication->getData('submissionId'));
+                $article = $article->getData('contextId') == $context->getId() ? $article : null;
```
This was only to ensure that the article is in the right context, but I think this is not needed because PubObjectsExportPlugin::getPublishedPublications() considers this... So I will remove...

bozana commented:
Hi @kaitlinnewson, thanks a lot for the review! I considered you latest comments. I actually did not know if you have finished the review, but... :-)
Thanks a lot!

kaitlinnewson commented on plugins/generic/doaj/templates/index.tpl:
```diff
@@ -0,0 +1,120 @@
+					{/foreach}
+				</div>
+			{/if}
+			<p><a href="http://www.doaj.org/application/new" target="_blank">{translate key="plugins.importexport.doaj.export.contact"}</a></p>
```
Not that I know of, other than linking to Learning OJS in the top right icons, which would not be as helpful in this case. Something to discuss further for these cases where we have specific docs for a plugin.

kaitlinnewson commented:
Hi @bozana, yes done reviewing this, thanks!

bozana commented:
Thanks a lot @kaitlinnewson! If no objections, I will rebase and create submodule updates for omp and ops, and once the tests pass I will merge... Tadaaaaa... :-)

kaitlinnewson commented on controllers/grid/submissions/ExportPublishedSubmissionsListGridHandler.php:
```diff
@@ -79,15 +79,18 @@ public function initialize($request, $args = null)
+        if (!$this->_plugin) {
+            // loadCategory because loadPlugin does not work properly when plugin name is provided:
+            // loadPlugin calls instantiatePlugin, that considers $pluginName to be plugin folder name and
+            // assumes class is $pluginName.'Plugin' (which is only the case for generic plugins) or
```
Looks like this comment may have been cut off

bozana commented on classes/submission/DAO.php:
```diff
@@ -79,11 +81,12 @@ public function getExportable(int $contextId, $pubIdType = null, $title = null,
-            ->where('s.status', '=', Submission::STATUS_PUBLISHED)
             ->where('s.context_id', '=', $contextId)
+            ->where('p.version_stage', '=', VersionStage::VERSION_OF_RECORD)
+            ->where('p.status', '=', Publication::STATUS_PUBLISHED)
```
@kaitlinnewson, for you a note, that I also made this change -- to consider only published VoR for Article export as well. Also, I do not check the submission status any more, only publication status. I think they already can diverge, or it will surely be so in the near future.
So maybe just for you to consider this when testing Zenodo.
Thanks!
